### ISSUE TITLE
vsce: design for new auth flow

### ISSUE THREAD

**Issue by abeatrix** [sourcegraph/sourcegraph#35899]
We have improved the new auth in https://github.com/sourcegraph/sourcegraph/pull/35339 which just added a button (`Continue with browser`) that would take the users to login via browser.
![image](https://user-images.githubusercontent.com/68532117/169901366-2e15e99a-c5a7-4b55-90e9-a2b1e8082b4d.png)

Would like to check with @jjinnii to see if more design work is required before we can release a new version with the new auth flow.

**Comment by jjinnii** [comment]
Thanks @abeatrix! Love how innovative this solution is 😄. A couple UX thoughts/questions that I have: 

1. Will our (especially enterprise) users feel secure with this UX? We're not really communicating what's happening on the Sourcegraph settings page when the user hits "Continue with browser". Do they feel as if their code and authentication are secure? 

   - Is the copy in this system notification customizable?   
    <img width="293" alt="Screen Shot 2022-05-24 at 2 25 55 PM" src="https://user-images.githubusercontent.com/87138876/169955456-515470eb-8ac5-4390-921c-e0517e659171.png">

   - Maybe it's worth testing with 1-2 enterprise users. @abeatrix do you remember the name of the person you worked with from Netflix to work through their auth problems back when we first released by any chance? I think it'd be awesome to follow up with them to say hey we improved our auth flow, wanna try it out?


3. What would happen if a user were to click "Import token to VS Code extension" from sourcegraph.com after they've followed these new steps? Is it potentially confusing that this action would still be in settings even after authenticating?

<img width="490" alt="Screen Shot 2022-05-24 at 2 28 59 PM" src="https://user-images.githubusercontent.com/87138876/169955913-aaff005a-f138-465b-ae59-58fe49342a6e.png">

**Comment by abeatrix** [comment]
@jjinnii Thank you ❤️ 


> Will our (especially enterprise) users feel secure with this UX? We're not really communicating what's happening on the Sourcegraph settings page when the user hits "Continue with browser". Do they feel as if their code and authentication are secure?

This is only available for Cloud users since only instances with the next release would have this feature. I have run through this with the Security team though and they said this is secured as we are not sending any data over the internet but locally (from the local browser back to local VS Code server). We might need a separate issue to make this feature available to Enterprise users since we will need to first check if their instances are at the latest version first. 

> Is the copy in this system notification customizable?

Unfortunately no as it comes from Chrome.

> What would happen if a user were to click "Import token to VS Code extension" from sourcegraph.com after they've followed these new steps? 

It will redirect them back to VS Code to import the token automatically. I added it there just in case the first time the pop up wasn't showing up for them or if they press `cancel` accidentally.


> Is it potentially confusing that this action would still be in settings even after authenticating?

I've added a message to express what the button does, but I'm not sure if the message is clear though. What do you think @jjinnii ?

![image](https://user-images.githubusercontent.com/68532117/170065884-4981507d-b065-47fb-afd7-d9dc271fc0ab.png)



## Screenshots

The video I shared was not up to date but here are a few screenshots of the latest build from main:
![Screen Shot 2022-05-24 at 7 25 33 AM](https://user-images.githubusercontent.com/68532117/170059853-97443678-1e36-49e6-a42e-15b58e236a27.png)
![image](https://user-images.githubusercontent.com/68532117/170064460-e79fa89b-ee69-468b-b741-1cdbc5e9d17b.png)

**Comment by jjinnii** [comment]
Hey @abeatrix! Took some time today to think through a new authentication flow for Cloud (and hopefully enterprise users). The designs are in this [Figma doc](https://www.figma.com/file/2RSo8aMS8KCBLOzk4Z5Cgq/VSCE-authentication-%26-feedback-flow-improvements-(Copy)?node-id=275%3A1628). There are two options that I've quickly mocked up: 

1. [Option A](https://www.figma.com/proto/2RSo8aMS8KCBLOzk4Z5Cgq/VSCE-authentication-%26-feedback-flow-improvements-(Copy)?page-id=246%3A11&node-id=2519%3A1533&viewport=615%2C-103%2C0.13&scaling=min-zoom&starting-point-node-id=2519%3A1533&show-proto-sidebar=1) is my recommended way, if technically feasible. An access token is the solution that we as Sourcegraph use to authenticate their account, but something not necessarily important to the user. An ideal flow doesn't expose this access token flow and automatically does this for them. 
2. [Option B ](https://www.figma.com/proto/2RSo8aMS8KCBLOzk4Z5Cgq/VSCE-authentication-%26-feedback-flow-improvements-(Copy)?page-id=246%3A11&node-id=2521%3A625&viewport=615%2C-103%2C0.13&scaling=min-zoom&starting-point-node-id=2521%3A670&show-proto-sidebar=1)is if Option A is not technically feasible, and we _need_ to expose the access token page. I wasn't sure how customizable this page was, but we want to reduce the noise as much as possible, while also communicating what is happening on this page. This is why I've added the step for the user to actually click "Import token" button so that they're consciously making the decision to do this. 

Let me know what your thoughts are on these two options and how technically feasible Option A might be!

**Comment by abeatrix** [comment]
Hi @jjinnii, not sure how to add comments in the figma links you shared so I will add my thoughts and comments here if that's ok 😃 

### Web App

1. Option A is also my preferred way. IT IS technically feasible, but I believe we should still redirect them to the new token callback page (between slide 6 and slide 7) because we won't be able to check if a pop up has shown up for the user (maybe their VS Code is closed, they canceled the popup by accident, or their browsers setting blocks popup, etc).
2. The page is very much customizable since we are the one who created it for integration purposes 😁 I have implemented your design and added a warning / informative message on the top to let the user know that a pop up dialogue is supposed to show up (see image below)

![Screen Shot 2022-05-26 at 11 28 21 AM](https://user-images.githubusercontent.com/68532117/170556159-e7eb7a7e-b0f2-41e6-8581-ed92e4968166.png)


### VSCE

I've implemented your design as shown in your figma files:

Step 1 - Choose if they have an account (go to step 2 when clicked) or if they need to create one (take them to browser, and it will redirect them to the token import page once they have completed sign up, so step 2 can be skipped)
![Screen Shot 2022-05-26 at 12 21 52 PM](https://user-images.githubusercontent.com/68532117/170561695-840f6b60-e799-43af-865d-0f3baf630f3c.png)

Step 2 - Choose user type: browser login for Cloud user (move to step 3a when clicked), or access token for private instance user (move to step 3b when clicked)
![Screen Shot 2022-05-26 at 11 26 05 AM](https://user-images.githubusercontent.com/68532117/170556354-7b4ffe2d-3abc-4484-b66f-aedf0e8d6d41.png)

Step 3a - Cloud user Login: access token panel will still be displayed in case users were not redirected back from web
![Screen Shot 2022-05-26 at 12 28 39 PM](https://user-images.githubusercontent.com/68532117/170563108-5b377c59-fe28-4349-af33-6625f650c8db.png)


Step 3b - Enterprise User Login: ask for instance URL and access token. Disable `Connect in Browser` button
![Screen Shot 2022-05-26 at 11 26 08 AM](https://user-images.githubusercontent.com/68532117/170559554-060dd2e0-4280-4d2a-928d-1bd4426d5fc4.png)



Please let me know your thoughts and your expertise on these!

**Comment by philipp-spiess** [comment]
@abeatrix is working on other things at the moment so we moved this ticket back into the backlog. What is the priority of working on these changes @ryankscott / @jjinnii ?

**Comment by ryankscott** [comment]
@philipp-spiess for this particular issue, it's low. It should be superseded by the two following issues:
1. https://github.com/sourcegraph/sourcegraph/issues/42074
2. https://github.com/sourcegraph/sourcegraph/issues/38072
